src/vibrato.rs

```diff
@@ -0,0 +1,232 @@
+use core::num;
+
+use crate::ring_buffer::RingBuffer;
+
+use crate::lfo::LFO;
+
+fn fclamp(x: f32, min_val: f32, max_val: f32) -> f32 {
+    if x < min_val {
+        min_val
+    } else if x > max_val {
+        max_val
+    } else {
+        x
+    }
+}
+macro_rules! assert_close {
+    ($left:expr, $right:expr, $epsilon:expr) => {{
+        let (left, right, epsilon) = ($left, $right, $epsilon);
+        assert!(
+            (left - right).abs() <= epsilon,
+            "{} is not close to {} within an epsilon of {}",
+            left,
+            right,
+            epsilon
+        );
+    }};
+}
+
+/// I chose a sample-by-sample processing scheme rather than an input block scheme
+/// because I think it is easier to use and much less complicated than a slice of a slice.
+/// It also would make it easier to use this object in a chain of effects.
+/// For the setters, I chose to have them separate rather than just having one because
+/// if you're using a modern editor, it is easier to type "myVibrato." to see the available methods than it is
+/// to type "myVibrato.set_param(vibrato::Params::Delay, value)" every time. 
+/// The core of my design is to make it easy to use this object without ever having to look at its source.
+pub struct Vibrato{
+    center_delay: f32,
+    max_delay: f32,
+    dry_wet: f32,
+    buffers: Vec<RingBuffer::<f32>>,
+    num_channels: usize,
+    sample_rate: f32,
+    lfo: LFO,
+    lfo_width: f32,
+}
+
+pub enum Params{
+    Delay,
+    Rate,
+    Width,
+    DryWet
+}
+
+impl Vibrato{
+    /// Creates a new Vibrato object
+    pub fn new(num_channels_: usize, sample_rate_: f32, max_delay_ms: f32)->Self{
+        let center_del = 0.005000  * sample_rate_; //  5 ms starting
+        let max_del = (max_delay_ms/1000.0) * sample_rate_;
+        let buffs_ = (0..num_channels_)
+            .map(|_| RingBuffer::<f32>::new(((max_delay_ms/1000.0)*sample_rate_+1.0) as usize))
+            .collect();
+        let lfo_ = LFO::new(sample_rate_, crate::lfo::WaveType::Sine
+                                    ,8.0, center_del* sample_rate_*1.0);
+         Vibrato{
+            center_delay:  center_del, // initialized at 5 ms
+            max_delay: max_del,
+            dry_wet: 1.0, // initialized at 100%
+            buffers: buffs_,
+            num_channels: num_channels_,
+            sample_rate: sample_rate_,
+            lfo: lfo_,
+            lfo_width: 1.0, 
+        }
+    }
+    
+    ///Takes input sample and channel id and 
+    /// outputs processed vibrato sample
+    pub fn process(&mut self, input_sample: f32, channel_id: usize)->f32{
+        assert!(channel_id < self.num_channels);
+        self.buffers[channel_id].push(input_sample);
+        let del = self.lfo.pop() + self.center_delay+1.0;
+        self.buffers[channel_id].pop_frac(del)*self.dry_wet + input_sample *(1.0-self.dry_wet)
```

Comment:
It seems like this line use residual value from the dry signal, which differed from the implementation in book, but makes sense since you can tune level of wet this way.